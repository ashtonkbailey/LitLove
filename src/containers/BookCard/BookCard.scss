.card {
  font-family: Sarabun;
  color: #ffffff;
  height: 360px;
  width: 330px;
  margin: 2em;
  overflow: scroll;
  padding: 1em;
  background: #363636;
  text-align: center;

  img {
    object-fit: none;
    position: relative;
    right: 1.8em;
  }

  h3 {
    margin-bottom: 0;
  }

  h4 {
    font-weight: normal;
    margin: 0;
  }

  button {
    background: #DAB192;
    box-shadow: 
      0px 4px 4px rgba(0, 0, 0, 0.25),
      inset -5px 6px 15px #EBCBB3;
    font-family: Sarabun;
    font-weight: bold;
    font-size: 12px;
    text-align: center;
    color: #363636;
    outline: none;
    cursor: pointer;
  }

  .confirm-btn {
    height: 2em;
    width: 5em;
    position: relative;
    bottom: 14.7em;
    right: 6em;
  }
}