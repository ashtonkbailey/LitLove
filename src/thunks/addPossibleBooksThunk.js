import { googleKey } from '../utils/googleAPI.js';
import { addPossibleBooks, setError } from '../actions';

export const addPossibleBooksThunk = (title) => {
  return async (dispatch) => {
    try {
      const url = `https://www.googleapis.com/books/v1/volumes?q=${title}&key=${googleKey}&maxResults=5&fields=items(id, volumeInfo/title, volumeInfo/authors, volumeInfo/description, volumeInfo/imageLinks)`;
      const response = await fetch(url);
      if (!response.ok) {
        throw Error('Unable to get book list')
      }
      const possibleBooks = await response.json();
      dispatch(addPossibleBooks(possibleBooks.items))
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}
