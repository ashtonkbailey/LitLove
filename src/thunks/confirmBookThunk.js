import { tDKey } from '../utils/tasteDiveAPI.js';
import { addLovedBook, setError } from '../actions';

export const confirmBookThunk = (bookTitle) => {
  return async (dispatch) => {
    try {
      const url = `https://cors-anywhere.herokuapp.com/https://tastedive.com/api/similar?q=${bookTitle}&type=books&info=1&k=${tDKey}`;
      const response = await fetch(url);
      if (!response.ok) {
        throw Error('Unable to get book recommendations')
      }
      const recommendedBooks = await response.json();
      console.log(recommendedBooks)
      dispatch(addLovedBook(bookTitle))
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}
