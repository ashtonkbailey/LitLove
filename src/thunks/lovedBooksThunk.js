import { googleKey } from '../utils/googleAPI.js';
import { addPossibleBooks } from '../actions';

export const lovedBookThunk = async (title) => {
  return async (dispatch) => {
    try {
      const url = `https://www.googleapis.com/books/v1/volumes?q=${title}&key=${googleKey}&fields=items(volumeInfo/title, volumeInfo/authors, volumeInfo/description, volumeInfo/imageLinks)`;
      const response = await fetch(url);
      if (!response.ok) {
        throw Error('Unable to get book list')
      }
      const possibleBooks = await response.json();
      dispatch(addPossibleBooks(possibleBooks))
    }
  }
}