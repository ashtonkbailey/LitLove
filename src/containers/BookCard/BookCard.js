import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

import { confirmBookThunk } from '../../thunks/confirmBookThunk'

class BookCard extends Component {
  constructor() {
    super();
    this.state = {
      confirmed: false
    }
  }
  handleClick = async (e) => {
    const bookTitle = this.props.volumeInfo.title;
    await this.props.confirmBook(bookTitle);
    this.setState({ confirmed: true })
  }

  render () {
    const { confirmed } = this.state;
    let bookTitle;
    let blurb;
    let fave;
    let imageURL;
    let author;

    if (this.props.type === 'confirm') {
      imageURL = this.props.volumeInfo.imageLinks.smallThumbnail;
      author = this.props.volumeInfo.authors[0];
      bookTitle = this.props.volumeInfo.title;
      blurb = this.props.volumeInfo.description;
      fave = ''
    } else {
      bookTitle = this.props.Name;
      blurb = this.props.wTeaser;
      fave = <button>star</button>
    }

    if (confirmed) {
      return <Redirect to="/recommendations" />
    }

    return (
      <div className="card">
        <button
          className="confirm-btn"
          onClick={this.handleClick}
        >Confirm</button>
        <img src={imageURL} alt={bookTitle} />
          {fave}
        <h3>{bookTitle}</h3>
        <h4>{author}</h4>
        <p>{blurb}</p>
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  confirmBook: (bookTitle) => dispatch(confirmBookThunk(bookTitle))
})

export default connect(null, mapDispatchToProps)(BookCard);
