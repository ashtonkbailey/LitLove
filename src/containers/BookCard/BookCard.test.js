import React from 'react';
import { shallow } from 'enzyme';

import BookCard, { mapDispatchToProps } from './BookCard';
import { confirmBookThunk } from '../../thunks/confirmBookThunk';

describe('BookCard', () => {
  let wrapper;
  let book;
  let bookId;
  let type;

  it('should match snapshot if type is confirm', () => {
    book = {
      "id": "ZI3gAQAACAAJ",
      "volumeInfo": {
        "title": "The Alchemist",
        "authors": [
          "Paulo Coelho"
        ],
        "description": "A special...",
        "imageLinks": {
          "smallThumbnail": "link",
          "thumbnail": "link"
        }
      }
    };
    bookId = book.id;
    type = 'confirm'
    wrapper = shallow(<BookCard {...book} key={bookId} type={type} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if type is recommendations', () => {
    book = {
      "Name": "Brida",
      "Type": "book",
      "wTeaser": "Paulo Coelho...",
      "wUrl": "http://en.wikipedia.org/wiki/Brida_(novel)",
      "yUrl": null,
      "yID": null
    };
    bookId = book.wUrl;
    type = 'recommendations'
    wrapper = shallow(<BookCard {...book} key={bookId} type={type} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleClick', () => {
    it.skip('should call confirmBook from props', () => {
      book = {
      "Name": "Brida",
      "Type": "book",
      "wTeaser": "Paulo Coelho...",
      "wUrl": "http://en.wikipedia.org/wiki/Brida_(novel)",
      "yUrl": null,
      "yID": null
      };
      bookId = book.wUrl;
      type = 'recommendations';
      const confirmBookMock = jest.fn();
      wrapper = shallow(<BookCard {...book} key={bookId} type={type} confirmBook={confirmBookMock}/>);
      wrapper.find('.confirm-btn').simulate('click');
      expect(wrapper.instance().confirmBook).toHaveBeenCalled()
    })
    
  });

  describe('mapDispatchToProps', () => {
    it.skip('should return a props object with the method confirmBookThunk', () => {
      const mockDispatch = jest.fn();
      confirmBookThunk.mockImplementation(() => 'Book confirmed');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.confirmBookThunk();
      expect(mockDispatch).toHaveBeenCalledWith('Book confirmed')
    })
  })
})
