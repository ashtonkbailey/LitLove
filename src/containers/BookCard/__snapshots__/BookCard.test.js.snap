// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard should match snapshot if type is confirm 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookCard
    id="ZI3gAQAACAAJ"
    type="confirm"
    volumeInfo={
      Object {
        "authors": Array [
          "Paulo Coelho",
        ],
        "description": "A special...",
        "imageLinks": Object {
          "smallThumbnail": "link",
          "thumbnail": "link",
        },
        "title": "The Alchemist",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <button
          className="confirm-btn"
          onClick={[Function]}
        >
          Confirm
        </button>
        <img
          alt="The Alchemist"
          src="link"
        />
        <h3>
          The Alchemist
        </h3>
        <h4>
          Paulo Coelho
        </h4>
        <p>
          A special...
        </p>
      </div>,
      "className": "card",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="confirm-btn"
            onClick={[Function]}
          >
            Confirm
          </button>,
          <img
            alt="The Alchemist"
            src="link"
          />,
          <h3>
            The Alchemist
          </h3>,
          <h4>
            Paulo Coelho
          </h4>,
          <p>
            A special...
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Confirm",
            "className": "confirm-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Confirm",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "The Alchemist",
            "src": "link",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "The Alchemist",
          },
          "ref": null,
          "rendered": "The Alchemist",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Paulo Coelho",
          },
          "ref": null,
          "rendered": "Paulo Coelho",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "A special...",
          },
          "ref": null,
          "rendered": "A special...",
          "type": "p",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <button
            className="confirm-btn"
            onClick={[Function]}
          >
            Confirm
          </button>
          <img
            alt="The Alchemist"
            src="link"
          />
          <h3>
            The Alchemist
          </h3>
          <h4>
            Paulo Coelho
          </h4>
          <p>
            A special...
          </p>
        </div>,
        "className": "card",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="confirm-btn"
              onClick={[Function]}
            >
              Confirm
            </button>,
            <img
              alt="The Alchemist"
              src="link"
            />,
            <h3>
              The Alchemist
            </h3>,
            <h4>
              Paulo Coelho
            </h4>,
            <p>
              A special...
            </p>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Confirm",
              "className": "confirm-btn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Confirm",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "The Alchemist",
              "src": "link",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "The Alchemist",
            },
            "ref": null,
            "rendered": "The Alchemist",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Paulo Coelho",
            },
            "ref": null,
            "rendered": "Paulo Coelho",
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "A special...",
            },
            "ref": null,
            "rendered": "A special...",
            "type": "p",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`BookCard should match snapshot if type is recommendations 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookCard
    Name="Brida"
    Type="book"
    type="recommendations"
    wTeaser="Paulo Coelho..."
    wUrl="http://en.wikipedia.org/wiki/Brida_(novel)"
    yID={null}
    yUrl={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <h3
          className="rec-title"
        >
          Brida
        </h3>
        <p>
          Paulo Coelho...
        </p>
      </div>,
      "className": "card",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="rec-title"
          >
            Brida
          </h3>,
          <p>
            Paulo Coelho...
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Brida",
            "className": "rec-title",
          },
          "ref": null,
          "rendered": "Brida",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Paulo Coelho...",
          },
          "ref": null,
          "rendered": "Paulo Coelho...",
          "type": "p",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <h3
            className="rec-title"
          >
            Brida
          </h3>
          <p>
            Paulo Coelho...
          </p>
        </div>,
        "className": "card",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="rec-title"
            >
              Brida
            </h3>,
            <p>
              Paulo Coelho...
            </p>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Brida",
              "className": "rec-title",
            },
            "ref": null,
            "rendered": "Brida",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Paulo Coelho...",
            },
            "ref": null,
            "rendered": "Paulo Coelho...",
            "type": "p",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
